File: arella-hr/frontend/src/lib/employee.ts
import api from "./api";

export interface Employee {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  phone: string | null;
  department: string | null;
  position: string | null;
  hire_date: string | null;
  salary_base: number | null;
  address: string | null;
  manager_id: number | null;
  manager_name: string | null;
  status: "active" | "inactive" | "on_leave";
  avatar_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface EmployeeCreate {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  department?: string;
  position?: string;
  hire_date?: string;
  salary_base?: number;
  address?: string;
  manager_id?: number;
  status?: "active" | "inactive" | "on_leave";
  avatar_url?: string;
}

export interface EmployeeUpdate {
  first_name?: string;
  last_name?: string;
  email?: string;
  phone?: string;
  department?: string;
  position?: string;
  hire_date?: string;
  salary_base?: number;
  address?: string;
  manager_id?: number;
  status?: "active" | "inactive" | "on_leave";
  avatar_url?: string;
}

export interface EmployeeListResponse {
  items: Employee[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface EmployeeListParams {
  page?: number;
  page_size?: number;
  search?: string;
  department?: string;
  status?: string;
  position?: string;
  include_inactive?: boolean;
}

export function listEmployees(params: EmployeeListParams = {}) {
  return api.get<EmployeeListResponse>("/employees", { params });
}

export function getEmployee(id: number) {
  return api.get<Employee>(`/employees/${id}`);
}

export function createEmployee(data: EmployeeCreate) {
  return api.post<Employee>("/employees", data);
}

export function updateEmployee(id: number, data: EmployeeUpdate) {
  return api.put<Employee>(`/employees/${id}`, data);
}

export function deactivateEmployee(id: number) {
  return api.patch<Employee>(`/employees/${id}/deactivate`);
}

export function restoreEmployee(id: number) {
  return api.patch<Employee>(`/employees/${id}/restore`);
}

export function deleteEmployee(id: number) {
  return api.delete(`/employees/${id}`);
}

/** Recursive org tree node (top-level employees + nested direct reports). */
export interface OrgNode extends Employee {
  manager?: OrgNode | null;
  reports: OrgNode[];
}

export function getOrgTree() {
  return api.get<OrgNode[]>("/employees/org-tree");
}

/**
 * A single payslip: a payroll entry enriched with its run's period and the
 * employee's identifying details (printable stub).
 */
export interface Payslip {
  entry_id: number;
  run_id: number;
  period_start: string;
  period_end: string;
  run_status: string;
  employee_id: number;
  employee_name: string;
  position: string | null;
  department: string | null;
  gross_salary: number;
  bonuses: number;
  deductions: number;
  net_pay: number;
  notes: string | null;
  generated_at: string | null;
}

/**
 * The employee record linked to the signed-in user (404 when unlinked).
 * Powers the employee self-service "My Home" page.
 */
export function getMyProfile() {
  return api.get<Employee>("/employees/me").then((r) => r.data);
}

/** An employee's pay history as payslips, newest period first. */
export function getPayslips(employeeId: number) {
  return api.get<Payslip[]>(`/employees/${employeeId}/payslips`).then((r) => r.data);
}

File: arella-hr/backend/app/schemas/employee.py
"""Pydantic schemas for employee resources."""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, EmailStr, Field


# ── request schemas ──────────────────────────────────────────────────────────


class EmployeeCreate(BaseModel):
    """Schema for creating a new employee."""

    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr
    phone: Optional[str] = Field(None, max_length=50)
    department: Optional[str] = Field(None, max_length=100)
    position: Optional[str] = Field(None, max_length=100)
    hire_date: Optional[date] = None
    salary_base: Optional[float] = Field(None, ge=0, le=9999999)
    address: Optional[str] = None
    manager_id: Optional[int] = None
    status: str = Field(default="active", pattern="^(active|inactive|on_leave)$")
    avatar_url: Optional[str] = None


class EmployeeUpdate(BaseModel):
    """Schema for updating an existing employee (all fields optional)."""

    first_name: Optional[str] = Field(None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(None, min_length=1, max_length=100)
    email: Optional[EmailStr] = None
    phone: Optional[str] = Field(None, max_length=50)
    department: Optional[str] = Field(None, max_length=100)
    position: Optional[str] = Field(None, max_length=100)
    hire_date: Optional[date] = None
    salary_base: Optional[float] = Field(None, ge=0, le=9999999)
    address: Optional[str] = None
    manager_id: Optional[int] = None
    status: Optional[str] = Field(None, pattern="^(active|inactive|on_leave)$")
    avatar_url: Optional[str] = None


# ── response schemas ─────────────────────────────────────────────────────────


class EmployeeOut(BaseModel):
    """Public employee response (no sensitive internal fields)."""

    id: int
    first_name: str
    last_name: str
    email: str
    phone: str | None
    department: str | None
    position: str | None
    hire_date: date | None
    salary_base: float | None
    address: str | None
    manager_id: int | None
    manager_name: str | None  # denormalised display name of the manager
    status: str
    avatar_url: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EmployeeOutWithManager(BaseModel):
    """Employee response with manager info nested (for org tree)."""

    id: int
    first_name: str
    last_name: str
    email: str
    phone: str | None
    department: str | None
    position: str | None
    hire_date: date | None
    salary_base: float | None
    address: str | None
    manager_id: int | None
    manager: "EmployeeOutWithManager | None" = None
    reports: list["EmployeeOutWithManager"] = []
    status: str
    avatar_url: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── query / list schemas ─────────────────────────────────────────────────────


class EmployeeListParams(BaseModel):
    """Query parameters for listing employees."""

    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    search: Optional[str] = None  # search across name & email
    department: Optional[str] = None
    status: Optional[str] = None  # active | inactive | on_leave
    position: Optional[str] = None


class PaginatedResponse(BaseModel):
    """Wrapper for paginated list responses."""

    items: list[EmployeeOut]
    total: int
    page: int
    page_size: int
    total_pages: int


# ── circular reference forward reference ─────────────────────────────────────


EmployeeOutWithManager.model_rebuild()